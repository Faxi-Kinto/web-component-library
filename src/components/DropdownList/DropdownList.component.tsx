import React, { useState, ChangeEvent } from 'react';
import Icon from '../Icon';
import * as Styled from './DropdownList.styles';
import { DropdownItem } from '.';
import { FieldProps } from '../../helpers/FieldProps';
import useField from '../../hooks/useField';

/**
 * @name DropdownList
 *
 * Dropdown list of selectable options
 *
 * @returns {JSX}
 */

type SelectOnChange = (event: React.ChangeEvent<HTMLSelectElement>) => void;

export type DropdownListProps = FieldProps<string, SelectOnChange> & {
  list: DropdownItem[];
  placeholder?: string;
  className?: string;
  description?: string;
  errstate?: boolean;
  label?: string;
  small?: boolean;
};

const DropdownList: React.FC<DropdownListProps> = (
  props: DropdownListProps
): JSX.Element => {
  const {
    list,
    onChange,
    value,
    error,
    placeholder,
    className,
    description,
    errstate,
    label,
    small,
  } = props;
  const [stateValue, setValue] = useState(value);

  const showError = useField(props);

  function handleOnChange(event: ChangeEvent<HTMLSelectElement>): void {
    if (onChange) onChange(event);
    else setValue(event.target.value);
  }

  return (
    <Styled.Container
      className={`dropdown${small ? ' dropdown--small' : ''}${
        className ? ' ' + className : ''
      }`}
    >
      {label && <label className="dropdown__label">{label}</label>}
      <div className="dropdown__select-wrapper">
        <select
          onChange={handleOnChange}
          value={value || stateValue || 'DEFAULT'}
          className={`dropdown__select-wrapper__select${
            showError || errstate
              ? 'dropdown__select-wrapper__select--error'
              : ''
          }`}
        >
          {placeholder ? (
            <option value="DEFAULT" disabled>
              {placeholder}
            </option>
          ) : (
            ''
          )}

          {list.map((item: any, index: number) => (
            <option key={index} value={item.value}>
              {item.label}
            </option>
          ))}
        </select>
        <Icon name="chevron-down" className="dropdown__icon" />
      </div>
      {description ? (
        <div className="dropdown__description-msg">{description}</div>
      ) : (
        ''
      )}
      {showError ? <div className="dropdown__error-msg">{error}</div> : ''}
    </Styled.Container>
  );
};

export default DropdownList;
