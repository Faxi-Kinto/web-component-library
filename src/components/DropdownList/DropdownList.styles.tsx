import styled from 'styled-components';
import { theme } from '../../config';
import {
  positionAbsolute,
  boxShadow,
  padding,
  size,
  pxToRem,
  fontSize,
} from '../../styles/basics';
import { marginChildren } from '../../styles/spacings';

export const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  position: relative;
  width: 100%;
  ${marginChildren(`0 0 ${pxToRem('5px')} 0`)};

  .dropdown {
    &__label {
      color: ${theme.pallet.GRAY};
      display: flex;
      height: ${pxToRem('24px')};
      align-items: flex-end;
      ${fontSize('16px')};
    }

    &__select-wrapper {
      position: relative;
      width: 100%;

      &__select {
        width: 100%;
        height: ${pxToRem('45px')};
        ${padding(pxToRem('8px'), pxToRem('16px'))};

        opacity: 1;
        outline: none;
        background-color: ${theme.pallet.WHITE};
        font-size: ${pxToRem('14px')};
        appearance: none;
        ${boxShadow('8px', '8px', '20px', theme.pallet.SHADOW)};
        border: ${pxToRem('0.5px')} solid ${theme.pallet.DISABLED};
        border-radius: 0;

        &.select--error {
          border: ${pxToRem('0.5px')} solid ${theme.pallet.ERROR};
        }

        &:focus,
        &.select--error {
          box-shadow: none;
        }
      }
    }

    &__icon {
      ${positionAbsolute('', pxToRem('28px'), '50%', '')};
      transform: translateY(50%);
      ${size('fit-content')};
      pointer-events: none;
      color: ${theme.pallet.LIGHT_BLUE_2};
    }

    &__description-msg,
    &__error-msg {
      flex-basis: 100%;
      color: ${theme.pallet.DISABLED};
      line-height: ${pxToRem('20px')};
    }

    &__error-msg {
      font-size: ${pxToRem('14px')};
      color: ${theme.pallet.ERROR};
    }
  }

  &.dropdown {
    &--small {
      .dropdown__icon {
        ${positionAbsolute('', pxToRem('12px'), '50%', '')};
        ${size('fit-content')};
        ${fontSize('10px')};
        color: ${theme.pallet.GRAY_BLUE};
      }
    }
  }
`;
