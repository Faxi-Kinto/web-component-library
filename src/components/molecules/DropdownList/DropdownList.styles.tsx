import styled from 'styled-components';
import { theme } from 'config';
import { positionAbsolute } from 'styles/basics';
import { pxToRem, fontSize } from 'styles/fonts';
import { marginChildren } from 'styles/spacings';

export const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  position: relative;
  width: 100%;
  ${marginChildren(`0 0 ${pxToRem('10px')} 0`)};

  > label {
    color: ${theme.pallet.GRAY};
    display: flex;
    height: 24px;
    align-items: flex-end;
    ${fontSize('16px')}
  }

  > select {
    width: 100%;
    height: ${pxToRem('45px')};
    padding: 0.5rem 1rem;

    opacity: 1;
    outline: none;
    background-color: ${theme.pallet.WHITE};
    font-size: ${pxToRem('14px')};
    appearance: none;
    box-shadow: 8px 8px 20px ${theme.pallet.SHADOW};
    border: 0.5px solid ${theme.pallet.DISABLED};
    border-radius: 0;

    &.select--error {
      border: 0.5px solid ${theme.pallet.ERROR};
    }

    &:focus,
    &.select--error {
      box-shadow: none;
    }
  }

  .__icon {
    ${positionAbsolute('50px', '25px', '', '')};
    pointer-events: none;
    color: ${theme.pallet.LIGHT_BLUE_2};
  }

  .dropdown__description-msg,
  .dropdown__error-msg {
    flex-basis: 100%;
    color: ${theme.pallet.DISABLED};
    line-height: ${pxToRem('20px')};
  }
  .dropdown__error-msg {
    font-size: ${pxToRem('14px')};
    color: ${theme.pallet.ERROR};
  }
`;
