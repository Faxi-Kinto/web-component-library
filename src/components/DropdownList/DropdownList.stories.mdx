import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';
import { useState } from 'react';

import DropdownList from './DropdownList.component';

export const listItems = [
  { value: 'test1', label: 'Test value 1' },
  { value: 'test2', label: 'Test value 2' },
  { value: 'test3', label: 'Test value 3' },
  { value: 'test4', label: 'Test value 4' },
  { value: 'test5', label: 'Test value 5' },
];
export const Custom = () => {
  const [value, setValue] = useState('');
  const dropdownHandler = e => {
    setValue(e.target.value);
  };
};

<Meta title="Core|Molecules/DropdownList" />

# DropdownList Component

Dropdown list of selectable options.

Implements `FormField API`, and it can be `embeded` with other atoms and molecules to construct more complex interfaces.

## Implementations

#### Default

<Preview>
  <Story name="default" decorators={[withKnobs]}>
    <DropdownList
      list={listItems}
      description={text('Field description', '')}
      error={text('Error message', 'This is Error message')}
      touched={boolean('Touched', false)}
      dirty={boolean('Dirty', false)}
      errstate={boolean('Embeded error override flag', false)}
    />
  </Story>
</Preview>

#### With placeholder

<Preview>
  <Story name="With Placeholder">
    <DropdownList
      placeholder="Choose your option"
      list={listItems}
      onSelect={Custom.dropdownHandler}
      value={Custom.value}
    />
  </Story>
</Preview>

#### Error

Component uses FormField `touched` and `dirty` to show error state and message.

Flag `errstate` is used only when component is embeded as part of some other component and you need to show error state without showing the defoult error message.

<Preview>
  <Story name="With Override Error state">
    <DropdownList list={listItems} errstate={true} />
  </Story>
</Preview>

## Props

<Props of={DropdownList} />

## Notes

```tsx
const listArray: DropdownList[] = [
  {value: 'test1', label: 'Test value 1'},
  {value: 'test2', label: 'Test value 2'},
  {value: 'test3', label: 'Test value 3'},
  {value: 'test4', label: 'Test value 4'},
  {value: 'test5', label: 'Test value 5'}
];

const [value, setValue] = useState('');

const handleDropdownSelect = ev => setValue(ev.target.value);

// You can use React standards value & onSelect props..
<DropdownList list={listArray} onSelect={handleDropdownSelect} value={value}/>

// or embed it directly in HTML5 form & use FormData to collect the value
<DropdownList list={listArray} />
```
