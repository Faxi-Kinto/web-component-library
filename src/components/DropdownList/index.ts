import DropdownList from './DropdownList.component';

export type DropdownItem = {
  value: string;
  label: string;
};

export default DropdownList;
