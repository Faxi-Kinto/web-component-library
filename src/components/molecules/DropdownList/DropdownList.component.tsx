import React, { useState, ChangeEvent } from 'react';
import { Icon } from 'components';
import * as Styled from './DropdownList.styles';
import { DropdownItem } from './';
import { FieldProps } from '../FormField/FieldProps';
import useField from 'hooks/useField';

/**
 * @name DropdownList
 *
 * Dropdown list of selectable options
 *
 * @returns {JSX}
 */

type SelectOnChange = (event: React.ChangeEvent<HTMLSelectElement>) => void;

export type DropdownListProps = FieldProps<string, SelectOnChange> & {
  list: DropdownItem[];
  placeholder?: string;
  className?: string;
  description?: string;
  errstate?: boolean;
  label?: string;
};

const DropdownList: React.FC<DropdownListProps> = (
  props: DropdownListProps
): JSX.Element => {
  const {
    list,
    onChange,
    value,
    error,
    placeholder,
    className,
    description,
    errstate,
    label,
  } = props;
  const [stateValue, setValue] = useState(value);

  const showError = useField(props);

  function handleOnChange(event: ChangeEvent<HTMLSelectElement>): void {
    if (onChange) onChange(event);
    else setValue(event.target.value);
  }

  return (
    <Styled.Container className={className}>
      {label && <label>{label}</label>}
      <select
        onChange={handleOnChange}
        value={value || stateValue || 'DEFAULT'}
        className={showError || errstate ? 'select--error' : ''}
      >
        {placeholder ? (
          <option value="DEFAULT" disabled>
            {placeholder}
          </option>
        ) : (
          ''
        )}

        {list.map((item, index) => (
          <option key={index} value={item.value}>
            {item.label}
          </option>
        ))}
      </select>
      <Icon name="chevron-down" className="__icon" />
      {description ? (
        <div className="dropdown__description-msg">{description}</div>
      ) : (
        ''
      )}
      {showError ? <div className="dropdown__error-msg">{error}</div> : ''}
    </Styled.Container>
  );
};

export default DropdownList;
