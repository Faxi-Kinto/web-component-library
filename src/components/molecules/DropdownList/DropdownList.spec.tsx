import React from 'react';
import renderer from 'react-test-renderer';
import DropdownList from './DropdownList.component';
import { DropdownItem } from './';

const listItems = [
  { value: 'test1', label: 'Test value 1' },
  { value: 'test2', label: 'Test value 2' },
  { value: 'test3', label: 'Test value 3' },
  { value: 'test4', label: 'Test value 4' },
  { value: 'test5', label: 'Test value 5' },
] as DropdownItem[];

test('DropdownList renders correctly', () => {
  const component = renderer.create(<DropdownList list={listItems} />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
