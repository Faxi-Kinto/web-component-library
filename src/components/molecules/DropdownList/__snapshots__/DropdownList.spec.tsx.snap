// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownList renders correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  position: relative;
  width: 100%;
}

.c0 > * {
  margin: 0 0 0.625rem 0;
}

.c0 > :last-child {
  margin: 0;
}

.c0 > label {
  color: #4E606A;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 24px;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  font-size: 1rem;
  line-height: 16px;
}

.c0 > select {
  width: 100%;
  height: 2.8125rem;
  padding: 0.5rem 1rem;
  opacity: 1;
  outline: none;
  background-color: #FFF;
  font-size: 0.875rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  box-shadow: 8px 8px 20px #00000029;
  border: 0.5px solid #95A1A6;
  border-radius: 0;
}

.c0 > select.select--error {
  border: 0.5px solid #EB4820;
}

.c0 > select:focus,
.c0 > select.select--error {
  box-shadow: none;
}

.c0 .__icon {
  position: absolute;
  top: 50px;
  right: 25px;
  pointer-events: none;
  color: #00a5b5;
}

.c0 .dropdown__description-msg,
.c0 .dropdown__error-msg {
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  color: #95A1A6;
  line-height: 1.25rem;
}

.c0 .dropdown__error-msg {
  font-size: 0.875rem;
  color: #EB4820;
}

<div
  className="c0"
>
  <select
    className=""
    onChange={[Function]}
    value="DEFAULT"
  >
    
    <option
      value="test1"
    >
      Test value 1
    </option>
    <option
      value="test2"
    >
      Test value 2
    </option>
    <option
      value="test3"
    >
      Test value 3
    </option>
    <option
      value="test4"
    >
      Test value 4
    </option>
    <option
      value="test5"
    >
      Test value 5
    </option>
  </select>
  <div
    className="c1 __icon"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-chevron-down fa-w-14 fa-1x "
      data-icon="chevron-down"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  
  
</div>
`;
